packages/react-components/react-dialog/src/components/Dialog/useDialog.ts

```diff
@@ -9,20 +15,177 @@ import type { DialogProps, DialogState } from './Dialog.types';
+  if (process.env.NODE_ENV === 'development') {
+    if (childrenArray.length !== 1 && childrenArray.length !== 2) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'Dialog must contain at least one child <DialogContent/>,\n' +
+          'and at most two children <DialogTrigger/> <DialogContent/> (in this order)',
+      );
+    }
+  }
+  switch (childrenArray.length) {
+    // case where there's a trigger followed by content
+    case 2:
+      return childrenArray as [trigger: React.ReactNode, content: React.ReactNode];
+    // case where there's only content
+    case 1:
+      return [undefined, childrenArray[0]];
+    // unknown case
+    default:
+      return [undefined, undefined];
+  }
+}
+
+/**
+ * Checks is click event is a proper Overlay click dismiss
+ */
+function isOverlayClickDismiss(event: React.MouseEvent, type: DialogModalType): boolean {
+  const isDefaultPrevented = normalizeDefaultPrevented(event);
+  return type === 'modal' && !isDefaultPrevented();
+}
+
+/**
+ * Checks if dialog is opening
+ */
+function isOpening(current: boolean, next: Extract<React.SetStateAction<boolean>, Function>) {
+  return !current && next(current) === true;
+}
+
+/**
+ * Focus first element on content when dialog is opened,
+ * in case there's no focusable element, then a eventlistener is added to document
+ * to ensure Escape keydown functionality
+ */
+function useFocusFirstElement({
+  open,
+  requestOpenChange,
+  modalType,
+}: Pick<DialogState, 'open' | 'requestOpenChange' | 'modalType'>) {
+  const { findFirstFocusable } = useFocusFinders();
+  const { targetDocument } = useFluent_unstable();
+  const contentRef = React.useRef<HTMLElement>(null);
+  const triggerRef = React.useRef<HTMLElement>(null);
+
+  React.useEffect(() => {
+    if (open) {
+      const element = contentRef.current && findFirstFocusable(contentRef.current);
+      if (element) {
+        element.focus();
+        // NOTE: if it's non-modal global listener to escape is necessary
+        if (modalType !== 'non-modal') {
+          return;
+        }
+      } else {
+        if (process.env.NODE_ENV === 'development') {
+          // eslint-disable-next-line no-console
+          console.warn('A Dialog should have at least one focusable element inside DialogContent');
+        }
+      }
+
+      if (triggerRef.current && targetDocument) {
+        const trigger = triggerRef.current;
+        if (targetDocument.activeElement === trigger) {
+          trigger.blur();
+        }
+        const listener = (event: KeyboardEvent) => {
+          if (isEscapeKeyDismiss(event, modalType)) {
+            requestOpenChange({
+              event,
+              open: false,
+              type: 'documentEscapeKeyDown',
+            });
+            trigger.focus();
+            event.stopImmediatePropagation();
+          }
+        };
+        targetDocument.addEventListener('keydown', listener, { passive: false });
```

Comment:
why do you use `passive` is it just for internal safety ?

Reply:
Just to be verbose about it! I like being explicit 😅